common-app-lib/src/cmr/common_app/api/request_logger.clj

```diff
@@ -0,0 +1,169 @@
+;; Ring Middleware Handlers
+
+(defn add-body-hashes
+  "Calculate the MD5 and SHA1 hash for the body of a response and include those
+   results in the response as HTTP headers. Headers are called Content-MD5 and
+   Content-SHA1. Content-MD5 is mentioned on the Wikipedia page:
+   https://en.wikipedia.org/wiki/List_of_HTTP_header_fields#Standard_request_fields
+   so it is presumed to be in common use."
+  [handler]
+  (fn [request]
+    (if-not (config/add-hash-headers)
+      (handler request)
+      (let [response (handler request)
+            ;; This is run after all responses
+            text (str (:body response))
+            body-md5 (digest/md5 text)
+            body-sha1 (digest/sha-1 text)
+            updated-response (-> response
+                                 (assoc-in [:headers "Content-MD5"] body-md5)
+                                 (assoc-in [:headers "Content-SHA1"] body-sha1))]
+        updated-response))))
+
+;; Action-logger should provide the same info as a standard NCSA Log
+;; ; 127.0.0.1 - - [2023-12-27 19:04:01.676] "GET /collections?keyword=any HTTP/1.1" 200 112 "-" "curl/8.1.2" 296
+
+;; Fields are as follows:
+;; ;
+;; IP Address
+;; login (ignored)
+;; -
+;; login (ignored)
+;; -
+;; [date-time]
+;; "
+;; HTTP VERB
+;; Relative URL
+;; "
+;; Status code
+;; response size
+;; "-"
+;; agent
+;; time
+
+(defn action-logger
+  "Log a request from Ring returning JSON data to be parsed by a log tool like
+   Splunk or ELK. This handler can be called multiple times at different positions
+   in the handlers call. When calling multiple times then pass in an ID (two
+   arity call) and the output will include this value in the log so that log
+   entries can be told appart. One may want to do this to see when a value logged
+   becomes valid and thus availible downstream."
+  ([handler]
+   (action-logger handler :ignore))
+  ([handler id]
+   (fn [request]
+     (if-not (config/custom-request-log)
+       (handler request)
+       (let [start (tk/now-ms)
+             now (dtp/clj-time->date-time-str (tk/now))
+             response (handler request)
+             query-params (params/assoc-query-params request "UTF-8")
+             form-params (params/assoc-form-params request "UTF-8")
+             ;; If adding or removing elements, change the log-version number so
+             ;; that reporting scripts can try to protect against change. Humans
+             ;; may be creating Splunk or ELK reports based on this content.
+             note (-> {"log-type" "action-log" "log-version" "1.0.0"}
+                      ;; As this handler can be called multiple times, if so,
+                      ;; include an id showing which instance is reporting this
+                      ;; information.
+                      (as-> data (if (= id :ignore) (assoc data "log-id" id) data))
+                      ;; assume that (add-body-hashes) has been run and reuse data
```

Comment:
if this func is tightly coupled or reliant on another I think we should mention that in the func description.

Reply:
That is not how I see it. I renamed the parameter and expanded out the help for that function. Some time ago another developer turned me on to the idea of not just passing in nil when you don't need something but to use a value to signify what you wanted, which is what I'm doing here. The first version of the function just defaults the value, otherwise you pass in an id. Does my change help?